File: examples/xor_example.js
/**
 * @file
 * @ingroup MLP_Internal_Logic
 */
#!/usr/bin/env node
/**
 * XOR problem example using facaded-mlp-cuda
 */

const { MLP, JsActivationType, JsOptimizerType } = require('..');

function main() {
  // Show available backends
  console.log('Available GPU backends:', MLP.availableBackends());
  console.log();

  // Create MLP for XOR problem with auto-selected backend
  const mlp = new MLP(2, [8], 1, {
    hiddenActivation: JsActivationType.Sigmoid,
    outputActivation: JsActivationType.Sigmoid,
    gpuBackend: 'auto',
  });

  // Set hyperparameters
  mlp.learningRate = 0.5;
  mlp.optimizer = JsOptimizerType.Adam;

  console.log(`Created model: ${mlp.toString()}`);
  console.log(`Using backend: ${mlp.gpuBackend}`);
  console.log();

  // XOR training data
  const X = [
    [0.0, 0.0],
    [0.0, 1.0],
    [1.0, 0.0],
    [1.0, 1.0],
  ];
  const y = [
    [0.0],
    [1.0],
    [1.0],
    [0.0],
  ];

  // Train
  console.log('Training...');
  const result = mlp.fit(X, y, 2000, true);

  console.log(`\nFinal loss: ${result.finalLoss.toFixed(6)}`);
  console.log();

  // Test predictions
  console.log('Predictions:');
  const predictions = mlp.predictBatch(X);
  for (let i = 0; i < X.length; i++) {
    console.log(
      `  Input: [${X[i]}] -> Target: ${y[i][0].toFixed(1)}, Prediction: ${predictions[i][0].toFixed(4)}`
    );
  }

  // Save model
  mlp.save('xor_model.json');
  console.log('\nModel saved to xor_model.json');

  // Load and test
  const mlp2 = MLP.load('xor_model.json');
  console.log(`Loaded model: ${mlp2.toString()}`);

  const testInput = [1.0, 0.0];
  const output = mlp2.predict(testInput);
  console.log(`\nTest prediction for [${testInput}]: ${output[0].toFixed(4)}`);

  // Feature importance
  const importance = mlp.featureImportance();
  console.log('\nFeature importance:');
  for (const { featureIndex, score } of importance) {
    console.log(`  Feature ${featureIndex}: ${score.toFixed(6)}`);
  }
}

main();

File: examples/xor_example.py
## @file
## @ingroup MLP_Internal_Logic
#!/usr/bin/env python3
"""
XOR problem example using facaded_mlp_cuda
"""

from facaded_mlp_cuda import MLP, PyActivationType, PyOptimizerType

def main():
    # Show available backends
    print("Available GPU backends:", MLP.available_backends())
    print()
    
    # Create MLP for XOR problem with auto-selected backend
    mlp = MLP(
        input_size=2,
        hidden_sizes=[8],
        output_size=1,
        hidden_activation=PyActivationType.Sigmoid,
        output_activation=PyActivationType.Sigmoid,
        gpu_backend="auto"  # or "cuda", "opencl", "cpu"
    )
    
    # Set hyperparameters
    mlp.learning_rate = 0.5
    mlp.optimizer = PyOptimizerType.Adam
    
    print(f"Created model: {mlp}")
    print(f"Using backend: {mlp.gpu_backend}")
    print()
    
    # XOR training data
    X = [
        [0.0, 0.0],
        [0.0, 1.0],
        [1.0, 0.0],
        [1.0, 1.0]
    ]
    y = [
        [0.0],
        [1.0],
        [1.0],
        [0.0]
    ]
    
    # Train
    print("Training...")
    losses = mlp.fit(X, y, epochs=2000, verbose=True)
    
    print(f"\nFinal loss: {losses[-1]:.6f}")
    print()
    
    # Test predictions
    print("Predictions:")
    predictions = mlp.predict_batch(X)
    for input_val, target_val, pred_val in zip(X, y, predictions):
        print(f"  Input: {input_val} -> Target: {target_val[0]:.1f}, Prediction: {pred_val[0]:.4f}")
    
    # Save model
    mlp.save("xor_model.json")
    print("\nModel saved to xor_model.json")
    
    # Load and test
    mlp2 = MLP.load("xor_model.json")
    print(f"Loaded model: {mlp2}")
    
    test_input = [1.0, 0.0]
    output = mlp2.predict(test_input)
    print(f"\nTest prediction for {test_input}: {output[0]:.4f}")
    
    # Feature importance
    importance = mlp.feature_importance()
    print("\nFeature importance:")
    for feature_idx, score in importance:
        print(f"  Feature {feature_idx}: {score:.6f}")

if __name__ == "__main__":
    main()
